File: src/mandelbrot.m
% Mandelbrot Set Calculation in Octave/MATLAB
% Computes the Mandelbrot set and measures execution time
%
% Run: octave mandelbrot.m

function result = mandelbrot(width, height, max_iter, xlim, ylim)
    if nargin < 1, width = 800; end
    if nargin < 2, height = 600; end
    if nargin < 3, max_iter = 100; end
    if nargin < 4, xlim = [-2.0, 1.0]; end
    if nargin < 5, ylim = [-1.0, 1.0]; end
    
    result = zeros(height, width);
    
    for py = 1:height
        y0 = ylim(1) + (py - 1) * (ylim(2) - ylim(1)) / (height - 1);
        for px = 1:width
            x0 = xlim(1) + (px - 1) * (xlim(2) - xlim(1)) / (width - 1);
            
            % Mandelbrot iteration
            x = 0.0;
            y = 0.0;
            iteration = 0;
            
            while (x*x + y*y <= 4.0 && iteration < max_iter)
                xtemp = x*x - y*y + x0;
                y = 2.0*x*y + y0;
                x = xtemp;
                iteration = iteration + 1;
            end
            
            result(py, px) = iteration;
        end
    end
end

% Main execution
tic;
result = mandelbrot();
result_sum = sum(result(:));
elapsed = toc;

fprintf('Result sum: %d\n', result_sum);
fprintf('Time: %f seconds\n', elapsed);

% Save result to CSV
fid = fopen('mandelbrot_octave_data.csv', 'w');
fprintf(fid, 'x,y,iter\n');

[height, width] = size(result);
for py = 1:height
    for px = 1:width
        fprintf(fid, '%d,%d,%d\n', px-1, py-1, result(py, px));
    end
end

fclose(fid);
fprintf('Result saved to mandelbrot_octave_data.csv\n');

File: src/mandelbrot.rs
// SPDX-License-Identifier: MIT
// Copyright (c) 2025 Mandelbrot Benchmark Project Contributors
//
// Mandelbrot Set Calculation in Rust
// Computes the Mandelbrot set and measures execution time
//
// Compile: rustc -C opt-level=3 mandelbrot.rs -o mandelbrot_rust.exe
// Run: .\mandelbrot_rust.exe

use std::time::Instant;
use std::fs::File;
use std::io::Write;

fn mandel_rust(width: usize, height: usize, max_iter: usize, 
               xlim: (f64, f64), ylim: (f64, f64)) -> Vec<Vec<usize>> {
    let mut result = vec![vec![0; width]; height];
    
    for py in 0..height {
        for px in 0..width {
            // Map pixel to complex plane
            let x0 = px as f64 * (xlim.1 - xlim.0) / width as f64 + xlim.0;
            let y0 = py as f64 * (ylim.1 - ylim.0) / height as f64 + ylim.0;
            
            // Mandelbrot iteration
            let mut x = 0.0;
            let mut y = 0.0;
            let mut iteration = 0;
            
            while x * x + y * y <= 4.0 && iteration < max_iter {
                let xtemp = x * x - y * y + x0;
                y = 2.0 * x * y + y0;
                x = xtemp;
                iteration += 1;
            }
            
            result[py][px] = iteration;
        }
    }
    
    result
}

fn main() {
    let width = 800;
    let height = 600;
    let max_iter = 100;
    let xlim = (-2.0, 1.0);
    let ylim = (-1.0, 1.0);
    
    let start = Instant::now();
    let result = mandel_rust(width, height, max_iter, xlim, ylim);
    let elapsed = start.elapsed();
    
    println!("Time: {:.6} seconds", elapsed.as_secs_f64());
    
    // Save result to CSV
    let mut file = File::create("mandelbrot_rust_data.csv").expect("Unable to create file");
    writeln!(file, "x,y,iter").expect("Unable to write header");
    
    for (py, row) in result.iter().enumerate() {
        for (px, &iter) in row.iter().enumerate() {
            writeln!(file, "{},{},{}", px, py, iter).expect("Unable to write data");
        }
    }
    
    println!("Result saved to mandelbrot_rust_data.csv");
}
